content/blog/2019-11-13-gep.md/index.md

```diff
@@ -0,0 +1,458 @@
++++
+title = "Making LLVM Address Calculation Safe(r)"
+extra.author = "Drew Zagieboylo"
+extra.bio = """
+  [Drew Zagieboylo](https://www.cs.cornell.edu/~dzag/) is a 3rd year PhD student researching Security, Hardware Design, and Programming Languages. He enjoys rock climbing and gaming in his free time.
+"""
++++
+
+# Memory Safety in LLVM
+
+LLVM IR code is not generally memory safe.
+While certain *obviously bad* behaviors are
+disallowed, it is not hard to write code that
+may execute out-of-bounds memory accesses at runtime.
+
+For instance, the size of an
+array may be statically known,
+but the access index may be unknown at compile time:
+
+```C
+int foo(int x) {
+    int tmp[10];
+    ...//some code to set values in tmp
+    return tmp[x];
+}
+```
+In this project we seek to improve the memory
+safety of LLVM programs by inserting dynamic bounds
+checks at runtime that cause the program to stop
+executing rather than violate memory safety.
+After running our compilation pass the aforementioned
+code would have the following runtime behavior:
+
+```C
+int foo(int x) {
+    int tmp[10];
+    ...//some code to set values in tmp
+    if (x >= 0 && x < 10) {
+       return tmp[x];
+    } else {
+       exit(1);
+    }
+}
+```
+### LLVM Address Calculation
+
+When compiling high level array and struct access to LLVM code, compilers
+generally use the [getelementptr](https://llvm.org/docs/GetElementPtr.html)
+(or GEP) instruction to calculate offsets into these memory allocations.
+GEP instructions have the nice property that they are type aware; offsets are phrased
+in terms of "number of elements" rather than "number of bytes."
+For example, the code in this stub dereferences some memory in
+the middle of a struct (specifically the last element of the `b` field).
+
+```C
+struct EX {
+  int a;
+  char b[3];
+  int *c;
+}
+
+struct X;
+...
+return X->b[2];
+```
+
+In LLVM, we could write a single GEP instruction to calculate
+the correct offset into the struct and then execute a `load` instruction
+to actually dereference the pointer.
+
+```C
+%1 = getelementptr %struct.EX,  %struct.EX* %X, i64 0, i32 1, i64 2
+%2 = load i8, i8* %1
+ret %2
+```
+
+This functionality makes GEP an ideal point for analyzing out-of-bounds accesses.
+Before a program might make an out-of-bounds access it has to
+acquire an out-of-bounds pointer. Usually, this means it
+executes a GEP whose result will then later be the argument of a
+load or store operation.
+
+Our approach in this implementation is to prevent the execution of
+any runtime GEP instructions that might lead to
+illegal memory accesses.
+*If a program can never acquire an out-of-bounds pointer,
+it can't violate memory safety.*
```

Comment:
This is certainly true, but there's an implicit connotation behind it that probably isn't true: that GEPs are the only way to acquire an out-of-bounds pointer. For example, you can [bitcast](https://llvm.org/docs/LangRef.html#bitcast-to-instruction) an arbitrary integer to a pointer that, when dereferenced directly without a GEP, would violate memory safety.

Reply:
Added note pointing to the "soundness and completeness" discussion on why this isn't a complete solution